### ISSUE TITLE
How to can I Reduce Validation Box Loss, Class Loss, and DFL Loss in Custom YOLOv8?

### ISSUE THREAD

**Issue by KeyzCoder** [ultralytics/ultralytics#21753]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I’m working on a custom YOLOv8 model with MBConv replacement on someblocks on backbone and inserting CBAM after the 2nd concat on the neck, aiming to enhance YOLO for better accuracy, efficiency, and generalization.
However, my validation box loss, class loss, and DFL loss are still high.
What are the most effective ways to lower these losses and further improve my results?

Any advice or suggestions would be greatly appreciated!

Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @KeyzCoder, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get moving quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Since you’re modifying the backbone and neck, please also share:
-Your model YAML and exact architecture changes (MBConv placement, CBAM insertion points)
-Training hyperparameters (optimizer, LR schedule, epochs, batch size, imgsz, augmentations)
-A few representative validation samples and label format examples
-Exact Ultralytics version and environment details
-Key excerpts of train/val logs showing the losses you’re concerned about

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks for the details. High val box/cls/DFL losses are usually driven by data/optimization rather than the backbone/neck edits. First establish a clean baseline on your dataset with an unmodified model (YOLO11 recommended) to isolate whether the custom changes are the cause, then compare losses/mAP:

```bash
yolo train model=yolo11n.pt data=path/to/data.yaml epochs=100 imgsz=640 plots=True
```

If the baseline is fine, focus on optimization rather than just scaling loss weights (changing `box/cls/dfl` only changes the printed loss, not model quality). Practical tweaks that typically lower these losses:
- Lower the initial LR and extend warmup; increase image size and batch if memory allows; close mosaic in the final epochs:
```bash
yolo train model=your_custom.yaml data=path/to/data.yaml epochs=300 imgsz=800 batch=-1 lr0=0.005 warmup_epochs=5 lrf=0.01 close_mosaic=10 plots=True
```
- Verify labels and class balance; visually audit a few hundred samples. After fixing any labels, delete `*.cache` inside your `labels/` folders so they rebuild.
- Keep the head/DFL unchanged if possible; many backbone/neck changes slow convergence. You can also try freezing early layers for stability, e.g., `freeze=10`.

For how YOLO computes and weights these losses, see the loss reference (box/cls/DFL and TaskAlignedAssigner) in the [loss docs](https://docs.ultralytics.com/reference/utils/loss/) and available training hyperparameters in the [Train docs](https://docs.ultralytics.com/modes/train/). General guidance on diagnosing training issues is in [Troubleshooting Common YOLO Issues](https://docs.ultralytics.com/guides/yolo-common-issues/).

If you’d like targeted advice, please share:
- Your exact `yolo train ...` command and `runs/.../results.png`
- A small sample of your dataset to reproduce
- The output of `yolo checks` and the training log summary

We’ll review and suggest dataset- and model-specific adjustments.

**Comment by KeyzCoder** [comment]
Hi, thank you for the suggestion earlier. I tried training again and here are my updated results:

### 📊 Training Results

**Enhanced YOLOv8 (300 epochs)**

```
epoch	time	train/box_loss	train/cls_loss	train/dfl_loss	metrics/precision(B)	metrics/recall(B)	metrics/mAP50(B)	metrics/mAP50-95(B)	val/box_loss	val/cls_loss	val/dfl_loss	lr/pg0	lr/pg1	lr/pg2
300	11564.1	1.53568	1.18879	1.69118	0.39687	0.34934	0.28912	0.10092	2.59029	2.96144	2.94126	1.90E-05	1.90E-05	1.90E-05
```

**Base YOLOv8 (100 epochs)**

```
epoch	time	train/box_loss	train/cls_loss	train/dfl_loss	metrics/precision(B)	metrics/recall(B)	metrics/mAP50(B)	metrics/mAP50-95(B)	val/box_loss	val/cls_loss	val/dfl_loss	lr/pg0	lr/pg1	lr/pg2
100	4564.07	1.47514	1.13295	1.46253	0.39652	0.30563	0.26323	0.08211	2.66091	3.08329	2.61645	2.84E-05	2.84E-05	2.84E-05
```

I noticed that with your advice, the **box loss is indeed reduced in the enhanced model**, although it took longer to train since I ran it for 300 epochs compared to 100 for the base model.

---

### ❓ Questions

1. Is it okay to compare the models at different epochs (100 vs. 300), or should I retrain both with the same number of epochs for a fairer comparison?

2. Do you have any tips on how I can clearly **explain how my custom model was made and integrated**? For example, I created new `.py` files for `CBAM` and `MBConv`, then called them inside the custom YAML definition. I’d like to document this process properly.

3. What is the **best available dataset** you would recommend for custom YOLOv8 training (to benchmark both the base and enhanced models more reliably)?

---

Thanks again for the guidance!

---

**Comment by glenn-jocher** [comment]
Thanks for sharing the results—your observation is valid. For a fair comparison, match the training schedule and settings for both runs (same epochs, LR schedule, imgsz, batch, augmentations) and compare best validation metrics (mAP50-95) rather than raw losses; you can also re-evaluate best checkpoints on the same split for an apples-to-apples view:
```bash
yolo val model=path/to/run_base/weights/best.pt data=path/to/data.yaml imgsz=640
yolo val model=path/to/run_enhanced/weights/best.pt data=path/to/data.yaml imgsz=640
```

To clearly document your integration, keep custom layers importable at build time and reference them in your YAML; a minimal, reproducible example helps others follow your steps:
```python
from ultralytics import YOLO
from my_layers import MBConv, CBAM  # ensure custom modules are in scope before model build

model = YOLO("custom.yaml")  # YAML uses MBConv/CBAM
model.train(data="data.yaml", epochs=300, imgsz=640, plots=True)
```
Explain what/why you changed, show the YAML diff, and include a short train/val script; see the YAML and training argument structure in the [Configuration docs](https://docs.ultralytics.com/usage/cfg/). If you plan a PR, please keep changes minimal, well-tested, and documented.

For benchmarking, MS COCO 2017 is the standard choice; for quick sanity checks use the built-in COCO8/COCO128 before scaling up:
```bash
yolo train model=enhanced.yaml data=coco8.yaml epochs=20 imgsz=640
```
Guidance on evaluation focus and setup is in [Insights on Model Evaluation and Fine-Tuning](https://docs.ultralytics.com/guides/model-evaluation-insights/), and small COCO variants are referenced in the [Hyperparameter Tuning Guide](https://docs.ultralytics.com/guides/hyperparameter-tuning/).

**Comment by KeyzCoder** [comment]
Thanks for the tips! I’ll retrain both models with the same settings/epochs and compare using mAP50-95 with yolo val on the best checkpoints. For the custom layers, I just made .py files for MBConv/CBAM and called them in the YAML—I’ll make sure they’re importable and document the changes properly. I’ll also test on COCO8/128 first before moving to COCO 2017. Appreciate the help!

**Comment by glenn-jocher** [comment]
Sounds good—match epochs/settings, fix the random seed and keep the exact train/val split; for evaluation use rectangular inference and save COCO JSON, and optionally report params/GFLOPs to contextualize architecture changes; see the [Validation docs](https://docs.ultralytics.com/modes/val/) for options.

```bash
# Reproducible training
yolo train model=enhanced.yaml data=path/to/data.yaml epochs=300 imgsz=640 seed=0

# Apples-to-apples validation
yolo val model=path/to/run_enhanced/weights/best.pt data=path/to/data.yaml imgsz=640 rect=True save_json=True
```

```python
# Model summary to compare params/GFLOPs
from ultralytics import YOLO
m = YOLO("enhanced.yaml")
m.info()
```
